### DISCUSSION TITLE
Overlapping Reconnection Attempts

### DISCUSSION THREAD

**Discussion by mikehudson2** [socketio/socket.io#4661] [Q&A]
In https://socket.io/docs/v4/client-options/#timeout  default 'timeout'=20,000 and default 'reconnectionDelay' = 5,000. If we use those defaults and set 'reconnectionDelayMax' = 5,000, does that mean that if the client is disconnected at t=0ms and stay disconnected for >15,000ms, then at t=15,000 until t=19,999 there will be 4 'live' reconnection attempts? If this is the case is that problematical?

**Comment by darrachequesne** [comment]
The `timeout` value is used when establishing the low-level connection to the server. For example, if the WebSocket connection cannot be established (but without triggering the `onerror()` handler), the connection attempt is cancelled and the next connection attempt is tried after `reconnectionDelay` ms.

So, in the worst case with `reconnectionDelay: 1000` and `randomizationFactor: 0`:

| t | event |
| --- | --- |
| 0 | 1st connection attempt |
| 20000 | connection attempt failure (after `timeout`) |
| 21000 | 2nd connection attempt (after `reconnectionDelay`) |
| 41000 | connection attempt failure (after `timeout`) |
| 43000 | 3rd connection attempt (after `reconnectionDelay * 2`) |
| 63000 | connection attempt failure |
...

Also, the reconnection delay increases over time, with a maximum value of `reconnectionDelayMax`.

**Reply by mikehudson2** [reply]
If the socket is connected, and the client subsequently disconnects (e.g. WiFi off) then we see a 'disconnect' event of reason 'ping timeout'. The client then immediately tries to reconnect and fails, generating a 'connect_error' followed by a 2nd reconnection attempt at 1,000, then a 3rd at 3,000 etc.. Each failed reconnection has its own 'connect_error'. The value of 'timeout' is irrelevant.  

Does this mean that there are two types of failed reconnections? 
i) Reconnection attempts which fail to establish a handshake with the server and thereby cause 'connect_errors'?; and
ii) Reconnections which successfully handshake with the server but do not receive a ping from it within a certain period?

Why is 'timeout' relevant in (ii) but not in (i)? In (ii) is 'timeout' effectively how long the client should wait for the first ping from the server? 

(reconnectionDelay:1000, reconnectionDelayMax: not set, randomizationFactor:0, timeout 20000)

**Reply by darrachequesne** [reply]
The `timeout` value is only relevant in a few cases:

- in HTTP long-polling mode, if the server handles the HTTP request but keeps it open without sending the handshake payload
- in WebSocket mode, if the server accepts the WebSocket connection but does not send the handshake payload

Once the handshake payload is received, the low-level connection is considered up and running, and the `timeout` value is not used anymore. (maybe `openTimeout` would be a better name for this option)

Source: https://github.com/socketio/socket.io-client/blob/89175d0481fc7633c12bb5b233dc3421f87860ef/lib/manager.ts#L354-L360
